src/ir/function-utils.h

```diff
@@ -43,6 +46,188 @@ inline bool equal(Function* left, Function* right) {
   return left->imported() && right->imported();
 }
 
+// Find which parameters are actually used in the function, that is, that the
+// values arriving in the parameter are read. This ignores values set in the
+// function, like this:
+//
+// function foo(x) {
+//   x = 10;
+//   bar(x); // read of a param index, but not the param value passed in.
+// }
+//
+// This is an actual use:
+//
+// function foo(x) {
+//   bar(x); // read of a param value
+// }
+inline std::unordered_set<Index> getUsedParams(Function* func) {
+  LocalGraph localGraph(func);
+
+  std::unordered_set<Index> usedParams;
+
+  for (auto& [get, sets] : localGraph.getSetses) {
+    if (!func->isParam(get->index)) {
+      continue;
+    }
+
+    for (auto* set : sets) {
+      // A nullptr value indicates there is no LocalSet* that sets the value,
+      // so it must be the parameter value.
+      if (!set) {
+        usedParams.insert(get->index);
+      }
+    }
+  }
+
+  return usedParams;
+}
+
+// Try to remove a parameter from a set of functions and replace it with a local
+// instead. This may not succeed if the parameter type cannot be used in a
+// local, or if we hit another limitation, in which case this returns false and
+// does nothing. If we succeed then the parameter is removed both from the
+// functions and from the calls to it, which are passed in (the caller must
+// ensure to pass in all relevant calls and call_refs).
+//
+// This does not check if removing the parameter would change the semantics
+// (say, if the parameter's value is used), which the caller is assumed to do.
+//
+// This assumes that the set of functions all have the same signature. The main
+// use cases are either to send a single function, or to send a set of functions
+// that all have the same heap type (and so if they all do not use some
+// parameter, it can be removed from them all).
+inline bool removeParameter(const std::vector<Function*> funcs,
```

Comment:
There are an awful lot of preconditions that have to be met to use this function correctly. I wonder if there is some more encapsulated abstraction that would work as well.

Reply:
Maybe the comment isn't clear enough - I think it's not that complex? Or maybe I don't understand your concern.

Another way to phrase the comment might be: This function assumes you have already checked for "logical" correctness. It then simply removes the parameter, and if you made a mistake then you would be breaking the wasm. The function does check for "hard constraints" and returns false if it hits them, such constraints are things that basically would make the wasm not validate if applied, or otherwise be obviously wrong.

Reply:
The preconditions:

 1. All calls and call_refs must be passed in.
 2. Removing the parameter should not change semantics.
 3. All functions must have the same signature.
 
I guess this isn't as bad as I initially thought, but this is more preconditions than I would expect from a general-purpose utility that's supposed to be useful in many potential situations. I guess it's more of a special-purpose utility that we only expect to ever call in a couple places.

Why can't we collapse the call sites into a single DAE optimization? It seems unfortunate to have such similar optimizations that share so much code but are still separate.

Reply:
Thanks, I see what you mean better now.

I think this is somewhat specialized and not fully general purpose, yes. But it is a way to avoid duplication between the existing DAE which optimizes single functions, and a future signature-pruning pass which does the same logical operation but on entire heap types at once. With this shared code split out, that new pass will be very simple.

We could in principle put all that code into a single file, DAE.cpp, and then these methods would go in there. Then it would be obvious it's not truly general purpose. I'd still prefer to have a new file SignaturePruning.cpp, however, just to keep that pass nice and separate and simple.

One option that might achieve both goals is to put this code in a separate file, but not in a "general" place like `ir/function-utils.h`? It could perhaps be in a file `remove-parameters.cpp` which would live alongside `DAE.cpp` and `SignaturePruning.cpp`?

Reply:
`remove-parameters.cpp` sounds good to me!